# 판다스
판다스는 파이썬에서 데이터 처리를 위해 존재하는 가장 인기 있는 라이브러리 입니다.

보통 2차원의 데이터를 효율적으로 가공/처리할 수 있는 다양하고 훌륭한 기능을 제공합니다.

넘파이보다 고수준의 API가 제공되어 있어서 많이 사용됩니다.

또한 판다스는 `csv`, `tab`와 같은 다양한 유형의 분리 문자로 column을 분리한 파일을 손십과 `DataFrame`으로 로딩할 수 있게 합니다.

판다스의 핵심 객체는 `DataFrame`입니다.
### DataFrame
`DataFrame`은 여러 개의 행과 열로 이루어진 2차원 데이터를 담는 구조체 입니다.

### Index
`Index`는 RDBMS의 Primary key처럼 개별 데이터를 고유하게 식별하는 Key값 입니다.

`Series`와 `DataFrame`은 모두 `Index`를 **Key**값으로 가지고 있습니다.

### Series
`Series`는 `DataFrame`과 다른 점이 column이 하나 뿐인 데이터 구체입니다.

반면 `DataFrame`은 column이 여러 개인 데이터 구조체 입니다.

결론 적으로 `DataFrame`은 여러개의 `Seires`로 이루어져 있습니다.

In [2]:
# 판다스 불러오기
import pandas as pd

## 판다스 사용하기 (1)
판다스는 다양한 포맷으로 된 파일을 `DataFrame`으로 로딩할 수 있는 편리한 API를 제공합니다.

`read_csv()`, `read_table()`, `read_fwf()`가 존재합니다.

이 중에서 `read_csv()`는 필드 구분 문자가 `,`이고 `read_table()`은 `\t`입니다.

### read_csv()

`read_csv()`를 사용할 경우 인자인 `sep`에 원하는 구분 문자를 입력하면 그 구분문자로 사용이 가능합니다.

ex) `read_csv('파일명', sep='\t')`로 하면 구분 문자가 탭이 됩니다.

또한, `read_csv()`와 `read_table()`은 딱히 기능상 큰 차이가 없기 때문에 둘중 아무거나 써도 무관합니다.


### read_fwf
**Fixed Width**로 고정 길이 기반의 칼럼 포맷을 `DataFrame`으로 로딩하기 위한 API입니다.

### Filepath
`read_csv(filepath_or_buffer, sep=',', ...)`이라고 한다면 가장 중요한 인자는 **filepath**입니다.

다른 인자는 지정하지 않으면 디폴트 값으로 할당되지만 

`Filepath`에는 로드하려는 데이터 파일의 경로를 포함한 파일명을 입력해야합니다.


In [4]:
titanic_df = pd.read_csv(r'./data/titanic_train.csv')
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [5]:
# 모든 데이터 출력하는 방법 -> 변수이름 입력
titanic_df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


`pd.read_csv()`를 하면 인자로 들어온 파일을 로딩해서 `DataFrame` 객체로 반환을 합니다.

또한, 제일 첫 번째 줄에 있던 데이터들이 Column명이 된 것을 확인할 수 있습니다.

그리고 제일 왼쪽은 Column명이 없는데 이곳이 바로 `Index`입니다.

모든 `DataFrame` 내의 데이터는 생성되는 순간 고유의 `Index`값을 가지게 됩니다.

In [14]:
# N개의 Row를 반환하는 법 -> Defalut는 5개입니다.
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


### DataFrame의 행과 열을 알아내는 법
`DataFrame`의 행과 열을 알아내는 가장 좋은 방법은 생성된 `DataFrame` 객체의 shape 변수를 이용하는 것입니다.

`shape` 변수는 `DataFrame`의 행과 열을 **튜플**형태로 반환합니다

In [7]:
print('DataFrame의 크기 : ',titanic_df.shape)

DataFrame의 크기 :  (891, 12)


### DataFrame의 정보 알기
`DataFrame`의 데이터 뿐만아닌 칼럼의 타입, NULL의 데이터 개수, 데이터 분포도 등의 메타데이터도 조회가 가능합니다.

이를 하기 위해서는 `info()`, `describe()`라는 메서드를 사용하면 됩니다.


In [10]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### info()
`info()`는 총 데이터 건수, 데이터 타입, NULL 건수에 대해서 알 수 있습니다.

In [9]:
titanic_df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### describe()

`describe()` 메서드는 칼럼별 숫자형 데이터값의

n-percentile 분포도, 평균값, 최대값, 최솟값 을 나타냅니다.

이 메서드는 오직 숫자형 칼럼의 분포도만 조사합니다. Object형은 출력에서 제외시킵니다.

이는 정확한 분포도를 알 수 없지만, 개략적인 분포도를 알 수 있습니다.

- `count`는 Not Null인 데이터 건수를 나타냅니다.
- `mean`은 전체 데이터의 평균값
- `std`는 표준편차
- `min`은 최솟값
- `max`는 최대값
- `25%`는 25 percentile 값
- `50%`는 50 percentile 값
- `75%`는 75 percentile 값

### 분석한 데이터 살펴보기

`PassengerID`는 승객 ID를 식별하는 칼럼이여서 1~891까지의 숫자가 차례로 할당되어서 분석을 위한 의미있는 속성이 아닙니다.

`Survived`는 min = 0이고 25~75% = 0, max = 1이므로 **0과 1로만** 이루어져있는 것을 볼 수 있습니다.

`Pclass`의 경우도 min = 1, 25~75% = 2 or 3, max = 3 입니다.

이러한 데이터는 숫자에 어떠한 의미가 존재하고 이는 **카테고리 칼럼**이라고 불립니다.

예를들어서 남자 = 1, 여자 = 0 같은 이러한 칼럼을 **카테고리 칼럼**이라고 합니다.

### Pclass 칼럼 분석하기
`DataFrame`의 `[]` 연산자 내부에 Column 명을 입력하면 `Serise`형태로 특정 칼럼 데이터 세트가 반환도비니다.

이런 반환한 `Series` 객체에 `value_counts()` 메서드를 호출하면 **해당 칼럼값의 유형과 건수**를 확인할 수 있습니다.

### value_counts
`value_counts()`는 지정된 칼럼의 데이터 값 건수를 많은 건수 순서대로 정렬하여 반환합니다.

`value_counts()`는 **`Serise`에서만 사용이 가능합니다.** `DataFrame`에서는 사용이 불가능 합니다.

In [11]:
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


### Pclass 칼럼 분석결과
`Pclass`는 3이 491개, 2가 184개, 1이 216개가 존재하는 것을 볼 수 있습니다.

결론적으로 `Pclass`는 1,2,3으로만 이루어져 있다는 것을 알 수 있습니다.

`DataFrame`의 `[]` 연사자 내부에 칼럼명을 입력하면 해당 `Series` 객체를 반환합니다!!

In [12]:
titanic_pclass = titanic_df['Pclass']
print(titanic_pclass)

0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64


다음처럼 하나의 `Pclass` Column만 반환되는 것을 볼 수 있습니다.

이제 `head()`를 사용하면 앞의 값만 추출 할 수 있습니다.

In [13]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

### head()로 반환한 값
`head()`로 반환 한 `Series`는 `Index`와 해당 데이터 값으로 반환이 됩니다.

왼쪽은 `Index`, 오른쪽은 데이터 값을 가집니다.

In [15]:
# data타입과 반환값 확인해보기
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64
